### PR TITLE
[LOG4J2-3415] Support JSON configuration using embedded JSON parser

### PR DESCRIPTION
**Pull request by jvz**
* Moves `JsonReader` from log4j-layout-json-template to log4j-api.
* Adds support for comments in parsing JSON for compatibility with existing Jackson usage.
* Adds new default `JsonConfigurationFactory` implementation using this.

### CODE REVIEW

**Comment by garydgregory** [comment]
Why does any parser need to be in the API module when all configurations
take place in the Core module? I don't think we should fatten up the API
with parsers.

Gary

**Comment by jvz** [comment]
For convenience with the PropertiesUtil enhancements Ralph proposed. Otherwise, it can certainly go in core instead.

**Comment by garydgregory** [comment]
> For convenience with the PropertiesUtil enhancements Ralph proposed. Otherwise, it can certainly go in core instead.

It seems like a bad idea to me to put a parser of any kind in a module that I would hope to be minimal in breadth and scope, as well as only focused on logging.

**Comment by jvz** [comment]
That's a good point. The `PropertiesUtil` enhancement will likely need some additional inversion of control or something to avoid having the JSON parsing done at that level. I'll move this to core, though I also need to revise the parser a bit as I think Jackson might also allow for duplicate keys (otherwise I'm not exactly sure how to represent a configuration for things like arbiters).

**Comment by rgoers** [comment]
@jvz You are correct. Accomplishing what I want to with Properties will be much simpler if the JsonReader is a utility class in the api module. I will absolutely need JSON support for that. But to be honest, I am not sure that PropertiesUtil really needs to be in the API. Or if we have one it could probably be greatly simplified since it controls only a very few system properties.

**Comment by ppkarwasz** [comment]
@rgoers: isn't the properties format enough for your properties enhancement? `log4j-core` just needs to provide a `PropertySource` for the JSON format.

IMHO `PropertiesUtil` should stay in the API. This way `log4j2.loggerContextFactory` (or its equivalent) does not need to be set using Java system properties. Ditto for the simple logger implementation.

**Comment by jvz** [comment]
I think one possibility would be to add an interface in log4j-api for loading properties and the general LoggerContext-specific properties. Then we can move the parsing logic to core.

We could also try to split up properties related to controlling the API versus controlling the Core.

**Comment by jvz** [comment]
I'm going to merge this as-is. We can figure out how to integrate JSON parsing with PropertiesUtil later.

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/JsonReaderTest.java [hunk 1]
```diff
@@ -375,6 +376,11 @@ public void test_nesting() {
                                 }})));
     }
 
+    @Test
```
Comment (vy) [inline log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/JsonReaderTest.java:379]: I'd prefer here a `@ParametrizedTest` touching different corner cases. Further, the following cases are not tested:
1. multi-line comment blocks (should pass)
1. not closed comment blocks (should fail)
1. not opened comment blocks (should fail)

Comment (jvz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/util/JsonReaderTest.java:379]: Added in https://github.com/apache/logging-log4j2/commit/e03f1832ed0791e5c489a65983de9ee707145ccd
